MySQL DB unknown table error

JDBC: MySQL component shows an error when `lookupRowByPrimaryKey`, `upsertRowByPrimaryKey` and `deleteRowByPrimaryKey` action functions are being used. Even though a connection to MySQL database is successful. The reported error has the following form:
```
Error: java.lang.RuntimeException: java.sql.SQLSyntaxErrorException: Unknown table 'test' in information_schema at io.elastic.jdbc.ColumnNamesWithPrimaryKeyProvider.getColumns(ColumnNamesWithPrimaryKeyProvider.java:95) at io.elastic.jdbc.ColumnNamesWithPrimaryKeyProvider.getMetaModel(ColumnNamesWithPrimaryKeyProvider.java:43) at io.elastic.sailor.ServiceMethods$2.execute(ServiceMethods.java:58) at io.elastic.sailor.Service.executeMethod(Service.java:99) at io.elastic.sailor.Service.getServiceInstanceAndExecute(Service.java:79) at io.elastic.sailor.Service.createServiceInstanceAndExecute(Service.java:70) at io.elastic.sailor.Service.main(Service.java:55) Caused by: java.sql.SQLSyntaxErrorException: Unknown table 'test' in information_schema at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:118) at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:95) at com.mysql.cj.jdbc.exceptions.SQLExceptionsMapping.translateException(SQLExceptionsMapping.java:122) at com.mysql.cj.jdbc.StatementImpl.executeQuery(StatementImpl.java:1247) at com.mysql.cj.jdbc.DatabaseMetaData$7.forEach(DatabaseMetaData.java:3149) at com.mysql.cj.jdbc.DatabaseMetaData$7.forEach(DatabaseMetaData.java:3137) at com.mysql.cj.jdbc.IterateBlock.doForAll(IterateBlock.java:56) at com.mysql.cj.jdbc.DatabaseMetaData.getPrimaryKeys(DatabaseMetaData.java:3190) at io.elastic.jdbc.ColumnNamesWithPrimaryKeyProvider.getColumns(ColumnNamesWithPrimaryKeyProvider.java:71) ... 6 more
```

This happens after the list of tables is shown.

Confirm that error present on the MySQL server version 5.6.
And the table was present and contained PK.

Creds name in the BW `MySQL "no table" issue`

Error listing below:
`

Error: java.lang.RuntimeException: java.sql.SQLSyntaxErrorException: Unknown table 'test' in information_schema at io.elastic.jdbc.ColumnNamesWithPrimaryKeyProvider.getColumns(ColumnNamesWithPrimaryKeyProvider.java:95) at io.elastic.jdbc.ColumnNamesWithPrimaryKeyProvider.getMetaModel(ColumnNamesWithPrimaryKeyProvider.java:43) at io.elastic.sailor.ServiceMethods$2.execute(ServiceMethods.java:58) at io.elastic.sailor.Service.executeMethod(Service.java:99) at io.elastic.sailor.Service.getServiceInstanceAndExecute(Service.java:79) at io.elastic.sailor.Service.createServiceInstanceAndExecute(Service.java:70) at io.elastic.sailor.Service.main(Service.java:55) Caused by: java.sql.SQLSyntaxErrorException: Unknown table 'test' in information_schema at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:118) at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:95) at com.mysql.cj.jdbc.exceptions.SQLExceptionsMapping.translateException(SQLExceptionsMapping.java:122) at com.mysql.cj.jdbc.StatementImpl.executeQuery(StatementImpl.java:1247) at com.mysql.cj.jdbc.DatabaseMetaData$7.forEach(DatabaseMetaData.java:3149) at com.mysql.cj.jdbc.DatabaseMetaData$7.forEach(DatabaseMetaData.java:3137) at com.mysql.cj.jdbc.IterateBlock.doForAll(IterateBlock.java:56) at com.mysql.cj.jdbc.DatabaseMetaData.getPrimaryKeys(DatabaseMetaData.java:3190) at io.elastic.jdbc.ColumnNamesWithPrimaryKeyProvider.getColumns(ColumnNamesWithPrimaryKeyProvider.java:71) ... 6 more
`

--- a/README.md
+++ b/README.md
@@ -1,6 +1,12 @@
 # JDBC-component
 ## Description
 This is an open source component for working with object-relational database management systems on [elastic.io platform](http://www.elastic.io "elastic.io platform").
+
+### Completeness Matrix
+![JDBC Component Completeness Matrix](https://user-images.githubusercontent.com/5710732/55414696-e7c7bf80-556b-11e9-8aa9-bfe2ebca4303.png)
+
+[JDBC Component Completeness Matrix](https://docs.google.com/spreadsheets/d/1sZr9ydJbMK8v-TguctmFDiqgjRKcrpbdj4CeFuZEkQU/edit?usp=sharing)
+
 ### Purpose
 With this component you will have following triggers:
 

--- a/README.md
+++ b/README.md
@@ -108,6 +108,8 @@ The format of ``Start Polling From (optional)`` field should be like ``yyyy-mm-d
 - ``ss`` - second
 - ``sss`` - millisecond (optional)
 
+*Note:* please see [**Current-limitations**](#current-limitations).
+
 ### SELECT trigger (Deprecated)
 This action exists in JDBC component only for backward compatibility. New [**Select trigger**](#select-trigger) is recommended to use.
 
@@ -215,9 +217,7 @@ This action exists in JDBC component only for backward compatibility. [**Upsert 
 - ``Oracle`` - compatible with Oracle Database 8.1.7 - 12.1.0.2
 - ``MSSQL`` - compatible with Microsoft SQL Server 2008 R2 and higher
 3. The current implementation of the action ``Upsert By Primary Key`` doesn't mark non-nullable fields as required fields at a dynamic metadata. In case of updating such fields with an empty value you will get SQL Exception ``Cannot insert the value NULL into...``. You should manually fill in all non-nullable fields with previous data, if you want to update part of columns in a row, even if data in that fields doesn't change. 
-
-## Known issues
-No known issues are there yet.
+4. **Get Rows Polling trigger** does not supported in Real-Time flows
 
 ## License
 Apache-2.0 © [elastic.io GmbH](https://www.elastic.io "elastic.io GmbH")

--- a/README.md
+++ b/README.md
@@ -108,7 +108,7 @@ The format of ``Start Polling From (optional)`` field should be like ``yyyy-mm-d
 - ``ss`` - second
 - ``sss`` - millisecond (optional)
 
-*Note:* please see [**Current-limitations**](#current-limitations).
+*Please Note: Component Snapshot will not be overwritten in Real-Time flows due to platform behaviour, so we strongly recommend to use Get Rows Polling trigger in Keen Flows only*
 
 ### SELECT trigger (Deprecated)
 This action exists in JDBC component only for backward compatibility. New [**Select trigger**](#select-trigger) is recommended to use.
@@ -217,7 +217,6 @@ This action exists in JDBC component only for backward compatibility. [**Upsert 
 - ``Oracle`` - compatible with Oracle Database 8.1.7 - 12.1.0.2
 - ``MSSQL`` - compatible with Microsoft SQL Server 2008 R2 and higher
 3. The current implementation of the action ``Upsert By Primary Key`` doesn't mark non-nullable fields as required fields at a dynamic metadata. In case of updating such fields with an empty value you will get SQL Exception ``Cannot insert the value NULL into...``. You should manually fill in all non-nullable fields with previous data, if you want to update part of columns in a row, even if data in that fields doesn't change. 
-4. **Get Rows Polling trigger** does not supported in Real-Time flows
 
 ## License
 Apache-2.0 © [elastic.io GmbH](https://www.elastic.io "elastic.io GmbH")

--- a/component.json
+++ b/component.json
@@ -1,5 +1,5 @@
 {
-  "title": "Database MySQL issue",
+  "title": "Database MySQL",
   "description": "Database JDBC connector",
   "credentials": {
     "verifier": "io.elastic.jdbc.JdbcCredentialsVerifier",

--- a/component.json
+++ b/component.json
@@ -1,5 +1,5 @@
 {
-  "title": "Database",
+  "title": "Database MySQL issue",
   "description": "Database JDBC connector",
   "credentials": {
     "verifier": "io.elastic.jdbc.JdbcCredentialsVerifier",

--- a/component.json
+++ b/component.json
@@ -1,5 +1,5 @@
 {
-  "title": "Database MySQL",
+  "title": "Database",
   "description": "Database JDBC connector",
   "credentials": {
     "verifier": "io.elastic.jdbc.JdbcCredentialsVerifier",

--- a/component.json
+++ b/component.json
@@ -105,8 +105,9 @@
     },
     "select": {
       "main": "io.elastic.jdbc.triggers.SelectTriggerOld",
-      "title": "SELECT (Deprecated)",
-      "description": "This trigger is already deprecated. Please use 'Select' trigger",
+      "title": "SELECT",
+      "deprecated": true,
+      "description": "DEPRECATED: Please use 'Select' Trigger",
       "type": "polling",
       "fields": {
         "tableName": {
@@ -206,7 +207,9 @@
     },
     "createOrUpdateRecord": {
       "main": "io.elastic.jdbc.actions.CreateOrUpdateRecord",
-      "title": "Create or update record (Deprecated)",
+      "title": "Create or update record",
+      "deprecated": true,
+      "description": "DEPRECATED: Please use 'Upsert Row By Primary Key' Action",
       "fields": {
         "tableName": {
           "viewClass": "SelectView",

--- a/src/main/java/io/elastic/jdbc/ColumnNamesProvider.java
+++ b/src/main/java/io/elastic/jdbc/ColumnNamesProvider.java
@@ -55,14 +55,18 @@ public class ColumnNamesProvider implements DynamicMetadataProvider, SelectModel
     JsonObjectBuilder properties = Json.createObjectBuilder();
     Connection connection = null;
     ResultSet rs = null;
+    String catalog = null;
     String schemaName = null;
     boolean isEmpty = true;
-    Boolean isOracle = configuration.getString("dbEngine").equals("oracle");
     Boolean isMssql = configuration.getString("dbEngine").equals("mssql");
+    Boolean isMysql = configuration.getString("dbEngine").equals("mysql");
     try {
       connection = Utils.getConnection(configuration);
       DatabaseMetaData dbMetaData = connection.getMetaData();
-      rs = dbMetaData.getPrimaryKeys(null, null, tableName);
+      if (isMysql) {
+        catalog = configuration.getString("databaseName");
+      }
+      rs = dbMetaData.getPrimaryKeys(catalog, null, tableName);
       if (tableName.contains(".")) {
         schemaName = tableName.split("\\.")[0];
         tableName = tableName.split("\\.")[1];

--- a/src/main/java/io/elastic/jdbc/ColumnNamesWithPrimaryKeyProvider.java
+++ b/src/main/java/io/elastic/jdbc/ColumnNamesWithPrimaryKeyProvider.java
@@ -57,9 +57,11 @@ public class ColumnNamesWithPrimaryKeyProvider implements DynamicMetadataProvide
     Connection connection = null;
     ResultSet rs = null;
     ResultSet rsPrimaryKeys = null;
+    String catalog = null;
     String schemaName = null;
     boolean isEmpty = true;
     Boolean isOracle = configuration.getString("dbEngine").equals("oracle");
+    Boolean isMysql = configuration.getString("dbEngine").equals("mysql");
     try {
       connection = Utils.getConnection(configuration);
       DatabaseMetaData dbMetaData = connection.getMetaData();
@@ -67,8 +69,11 @@ public class ColumnNamesWithPrimaryKeyProvider implements DynamicMetadataProvide
         schemaName = tableName.split("\\.")[0];
         tableName = tableName.split("\\.")[1];
       }
+      if (isMysql) {
+        catalog = configuration.getString("databaseName");
+      }
       rsPrimaryKeys = dbMetaData
-          .getPrimaryKeys(null, ((isOracle && !schemaName.isEmpty()) ? schemaName : null),
+          .getPrimaryKeys(catalog, ((isOracle && !schemaName.isEmpty()) ? schemaName : null),
               tableName);
       rs = dbMetaData.getColumns(null, schemaName, tableName, "%");
       while (rs.next()) {

--- a/src/main/java/io/elastic/jdbc/PrimaryColumnNamesProvider.java
+++ b/src/main/java/io/elastic/jdbc/PrimaryColumnNamesProvider.java
@@ -55,14 +55,19 @@ public class PrimaryColumnNamesProvider implements DynamicMetadataProvider, Sele
     JsonObject properties = Json.createObjectBuilder().build();
     Connection connection = null;
     ResultSet rs = null;
+    String catalog = null;
     String schemaName = null;
     Boolean isEmpty = true;
     Boolean isOracle = configuration.getString("dbEngine").equals("oracle");
     Boolean isMssql = configuration.getString("dbEngine").equals("mssql");
+    Boolean isMysql = configuration.getString("dbEngine").equals("mysql");
     List<String> primaryKeys = new ArrayList();
     try {
       connection = Utils.getConnection(configuration);
       DatabaseMetaData dbMetaData = connection.getMetaData();
+      if (isMysql) {
+        catalog = configuration.getString("databaseName");
+      }
       if (tableName.contains(".")) {
         schemaName =
             (isOracle) ? tableName.split("\\.")[0].toUpperCase() : tableName.split("\\.")[0];
@@ -70,7 +75,7 @@ public class PrimaryColumnNamesProvider implements DynamicMetadataProvider, Sele
             (isOracle) ? tableName.split("\\.")[1].toUpperCase() : tableName.split("\\.")[1];
       }
       rs = dbMetaData
-          .getPrimaryKeys(null, ((isOracle && !schemaName.isEmpty()) ? schemaName : null),
+          .getPrimaryKeys(catalog, ((isOracle && !schemaName.isEmpty()) ? schemaName : null),
               tableName);
       while (rs.next()) {
         primaryKeys.add(rs.getString("COLUMN_NAME"));

--- a/src/main/java/io/elastic/jdbc/Utils.java
+++ b/src/main/java/io/elastic/jdbc/Utils.java
@@ -51,6 +51,7 @@ public class Utils {
     try {
       return DriverManager.getConnection(connectionString, user, password);
     } catch (Exception e) {
+      LOGGER.error("Failed while connecting. Error: " + e.getMessage());
       throw new RuntimeException(e);
     }
   }

--- a/src/main/java/io/elastic/jdbc/Utils.java
+++ b/src/main/java/io/elastic/jdbc/Utils.java
@@ -51,7 +51,7 @@ public class Utils {
     try {
       return DriverManager.getConnection(connectionString, user, password);
     } catch (Exception e) {
-      LOGGER.info("Failed while connecting");
+      LOGGER.info("Failed while connecting. Error: " + e.getMessage());
       throw new RuntimeException(e);
     }
   }

--- a/src/main/java/io/elastic/jdbc/Utils.java
+++ b/src/main/java/io/elastic/jdbc/Utils.java
@@ -51,6 +51,7 @@ public class Utils {
     try {
       return DriverManager.getConnection(connectionString, user, password);
     } catch (Exception e) {
+      LOGGER.info("Failed while connecting");
       throw new RuntimeException(e);
     }
   }

--- a/src/main/java/io/elastic/jdbc/Utils.java
+++ b/src/main/java/io/elastic/jdbc/Utils.java
@@ -51,7 +51,7 @@ public class Utils {
     try {
       return DriverManager.getConnection(connectionString, user, password);
     } catch (Exception e) {
-      LOGGER.info("Failed while connecting. Error: " + e.getMessage());
+      LOGGER.error("Failed while connecting. Error: " + e.getMessage());
       throw new RuntimeException(e);
     }
   }

--- a/src/main/java/io/elastic/jdbc/actions/UpsertRowByPrimaryKey.java
+++ b/src/main/java/io/elastic/jdbc/actions/UpsertRowByPrimaryKey.java
@@ -30,6 +30,7 @@ public class UpsertRowByPrimaryKey implements Module {
     JsonObject resultRow;
     String tableName;
     String dbEngine;
+    String catalog = null;
     String schemaName = "";
     String primaryKey = "";
     int primaryKeysCount = 0;
@@ -54,9 +55,13 @@ public class UpsertRowByPrimaryKey implements Module {
 
     LOGGER.info("Executing lookup primary key");
     boolean isOracle = dbEngine.equals(Engines.ORACLE.name().toLowerCase());
+    Boolean isMysql = configuration.getString("dbEngine").equals("mysql");
 
     try (Connection connection = Utils.getConnection(configuration)) {
       DatabaseMetaData dbMetaData = connection.getMetaData();
+      if (isMysql) {
+        catalog = configuration.getString("databaseName");
+      }
       if (tableName.contains(".")) {
         schemaName =
             (isOracle) ? tableName.split("\\.")[0].toUpperCase() : tableName.split("\\.")[0];
@@ -64,7 +69,7 @@ public class UpsertRowByPrimaryKey implements Module {
             (isOracle) ? tableName.split("\\.")[1].toUpperCase() : tableName.split("\\.")[1];
       }
       try (ResultSet rs = dbMetaData
-          .getPrimaryKeys(null, ((isOracle && !schemaName.isEmpty()) ? schemaName : null),
+          .getPrimaryKeys(catalog, ((isOracle && !schemaName.isEmpty()) ? schemaName : null),
               tableName)) {
         while (rs.next()) {
           primaryKey = rs.getString("COLUMN_NAME");

